### ISSUE TITLE
Support for "equation" environment

### ISSUE THREAD

**Issue by invkrh** [KaTeX/KaTeX#445]
It seems that multi-line auto rendering is not possible for the moment.

For example:

```
\begin{equation}
  x = a_0 + \cfrac{1}{a_1
          + \cfrac{1}{a_2
          + \cfrac{1}{a_3 + \cfrac{1}{a_4} } } }
\end{equation}
```

Here is the exception:

```
KaTeX auto-render: Failed to parse `
    \begin{equation}
    x = a_0 + \cfrac{1}{a_1
            + \cfrac{1}{a_2
            + \cfrac{1}{a_3 + \cfrac{1}{a_4} } } }
  \end{equation}\frac{1}{2} 
    ` with  Error: KaTeX parse error: No such environment: equation at position 21: begin{equation}̲
```

Any ideas ?

**Comment by invkrh** [comment]
The equation env has not been implemented yet...

But the `aligned*` is implemented according to the wiki page: https://github.com/Khan/KaTeX/wiki/Function-Support-in-KaTeX

It does not work neither.

```
\begin{aligned}
a&=b&c&=d\\
e&=f
\end{aligned}
```
It gives `No such environment: aligned*` exception

**Comment by invkrh** [comment]
It seems that `aligned` is not recognized by the environment.
https://github.com/Khan/KaTeX/blob/master/src/Parser.js#L428

**Comment by kevinbarabash** [comment]
`aligned` is supported on master but not in the most recent release.  I can see how the wiki page is a little confusing.  The `*` after `aligned` is supposed indicate that there is a note at the bottom of the wiki page that states `*Committed to master but not part of the most current release.`

**Comment by xymostech** [comment]
(I replaced `*` with `†` to make things less confusing maybe)

**Comment by invkrh** [comment]
Thank you for the clarification. =)

**Comment by edigleyalexandre** [comment]
@invkrh I also hope for the implementation of **\begin {equation}** for Katex. It would be fantastic.

**Comment by edemaine** [comment]
It is possible to add very basic support for `\begin{equation} ... \end{equation}` by specifying a custom delimiter in auto-render. For example:
```js
renderMathInElement(document.body, {
  delimiters: [
    {left: "$$", right: "$$", display: true},
    {left: "\\(", right: "\\)", display: false},
    {left: "\\[", right: "\\]", display: true},
    {left: "\\begin{equation}", right: "\\end{equation}", display: true}
  ]
});
```
See https://codepen.io/edemaine/pen/EpNamj for a demo.

However, equation numbering isn't supported yet; #350 tracks that.  Note that manual numbering via `\tag` is possible, though.

**Comment by edigleyalexandre** [comment]
@edemaine And what about a dollar sign?

**Comment by edemaine** [comment]
@edigleyalexandre You can add support for `$` by adding another delimiter to the list.

**Comment by ronkok** [comment]
@edigleyalexandre It matters where in the list things are placed. Be sure that "$" comes after "$$", as in:

```js
renderMathInElement(document.body, {
  delimiters: [
    {left: "$$", right: "$$", display: true},
    {left: "$", right: "$", display: true},
    {left: "\\(", right: "\\)", display: false},
    {left: "\\[", right: "\\]", display: true},
    {left: "\\begin{equation}", right: "\\end{equation}", display: true}
  ]
});
```

**Comment by edigleyalexandre** [comment]
I added as I said, but it does not render. Look [here](https://latex.obaricentrodamente.com/2018/01/exemplos-em-latex.html).

**Comment by edigleyalexandre** [comment]
@edemaine Here's how to get with Mathjax. On [here](https://latex.obaricentrodamente.com/2018/01/exemplos-em-latex.html).

**Comment by edigleyalexandre** [comment]
`\begin {matrix}` and `\begin {cases}` are supported on Katex. I tested the main [page](https://khan.github.io/KaTeX/) of the Katex website and rendered it, but the blog did not render. What happened? [Page](https://www.prof-edigleyalexandre.com/2018/03/calculadora-para-sistemas-lineares-com-ate-4-equacoes.html) with `\begin {cases}` that did not render.

**Comment by ronkok** [comment]
Well, you did omit a comma that should follow ` {left: "\(", right: "\)", display: false}`. I don't know if that by itself is enough to stop the script.

**Comment by edigleyalexandre** [comment]
It worked. Forgive me for that.

**Comment by N0rbert** [comment]
This is still the problem. It is impossible to render *equation*-based environments such as

```
Environment `\begin{equation}...\end{equation}`
with auto-numbering:

\begin{equation}
h(z) = \sinh(z)
\end{equation}

---

Environment `\begin{equation}...(\#eq:label)\end{equation}`
with auto-numbering and label (for bookdown):

\begin{equation}
  f\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
  (\#eq:binom)
\end{equation}

---

Environment `\begin{equation}...\label{label}\end{equation}`
with auto-numbering label for xaringan

\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

in several Markdown editors - see at least https://github.com/benweet/stackedit/issues/1424  and https://github.com/laurent22/joplin/issues/2212 .

**Comment by Naegolus** [comment]
omg. This is awesome!

**Comment by N0rbert** [comment]
Looks like it is not fixed, for example Joplin do not show `\begin{equation}...\end{equation}` blocks.

**Comment by zhangchenkai** [comment]
`\begin{equation}...\end{equation}` is so common in LaTex, why not consider to support this ?

**Comment by ronkok** [comment]
Closed via PR#2369.

**Comment by Varun-msd** [comment]
<img width="379" alt="Screenshot 2020-10-28 at 3 58 02 PM" src="https://user-images.githubusercontent.com/71187534/97424214-677b6300-1936-11eb-8087-e4bf56304a31.png">
`Without equation, it is working as overflow: scroll

$Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)$

Without array it is working as expected (Breaking properly)
\begin{equation}
Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)
\end{equation}

With array and without equation (Not Breaking properly, overflow is scrolling)
$3Ta{n^{ - 1}}x$$=\left\{\begin{array}{ccc}
Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)
 \end{array}\right\}$

With array and with equation it is not working as expected (Curly braces has problem) - **THIS IS MY ISSUE**
$3Ta{n^{ - 1}}x$$=\left\{\begin{array}{ccc} \begin{equation}
Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right)Ta{n^{ - 1}}\left( {\cfrac{{3x - {x^3}}}{{1 - 3{x^2}}}} \right) \end{equation}
 \end{array}\right\}$
`

ISSUE: **Why if I add an equation inside an array it's breaking**?

**Comment by ronkok** [comment]
@Varun-msd Soft line breaks occur after **top-level** relations and binary operators. If you wrap an `equation` environment or an `array` environment around some math, then the nested operators are no longer top-level.

Why was it programmed to work this way? Because KaTeX emulates TeX and that is the way TeX does soft line breaks. It is so specified in *The TeXbook* on page 173.

**Comment by Varun-msd** [comment]
So @ronkok you say that we can't achieve the solution of my problem in Katex ? , my question is because it works in MathJax 2.0 but not in MathJax 3.0 (as they don't support line breaking till now)

**Comment by N0rbert** [comment]
The bug is not fully fixed. The following syntax is not rendered:

```
\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

See below:

![label-fail](https://user-images.githubusercontent.com/309456/110846275-a8a85a00-82bc-11eb-9c37-2bafaaa3a97d.png)

Please: a - reopen, b - ask me to open new bug.

**Comment by edemaine** [comment]
@N0rbert `\label` is different functionality, and already tracked in different issues, most recently #2798.

**Comment by gnuion** [comment]
you can use align instead of flalifn https://katex.org/docs/supported.html#environments
```math
\begin{align}
(a+b+c+..)^m&=\sum_{i}a^i(b+c+..)^{m-i} && \text{(nach No. 25)} \\
&=\sum_{i}a^i(\sum\limits_{j}b^j(c+..)^{m-i-j}) && \text{(nach No. 25)} \\
&=\sum_{i,j}a^ib^j(c+..)^{m-(i+j)} && \text{(nach No. 2)} \\
&=\sum_{i,j,k..}a^ib^jc^k+..)^{m-(i+j+k..)} \text{ wo, } i+j+k+..=m && \text{(nach No. 25)} \\
\end{align}
```
